src/analyses/dependence_graph.cpp

```diff
@@ -26,37 +26,23 @@ bool dep_graph_domaint::merge(
   goto_programt::const_targett from,
   goto_programt::const_targett to)
 {
-  bool changed=has_values.is_false();
-  has_values=tvt::unknown();
+  bool changed = is_bottom() || has_changed;
 
-  depst::iterator it=control_deps.begin();
-  for(const auto &c_dep : src.control_deps)
+  if(is_bottom())
   {
-    while(it!=control_deps.end() && *it<c_dep)
-      ++it;
-    if(it==control_deps.end() || c_dep<*it)
-    {
-      control_deps.insert(it, c_dep);
-      changed=true;
-    }
-    else if(it!=control_deps.end())
-      ++it;
+    has_values = tvt::unknown();
   }
 
-  it=data_deps.begin();
-  for(const auto &d_dep : src.data_deps)
+  if(is_bottom() || (has_changed && data_deps.empty()))
   {
-    while(it!=data_deps.end() && *it<d_dep)
-      ++it;
-    if(it==data_deps.end() || d_dep<*it)
-    {
-      data_deps.insert(it, d_dep);
-      changed=true;
-    }
-    else if(it!=data_deps.end())
-      ++it;
+    data_deps = src.data_deps;
```

Comment:
Explain why when our own data deps are non-empty, we don't use src's data deps at all?

Reply:
We wouldn't actually need a fixpoint computation for the data dependency analysis as it merely uses the result of the prior reaching definitions analysis to compute the data dependencies at each location. For this we need to inspect each location only once. Thus if the set `data_deps` is non-empty it means we have handled that location already.

That's why it may be a good to split up the data dependency and control dependency analysis (see issue #1859), and then implement the data dependency analysis without `ait` by simply iterating over all instructions.